### DISCUSSION TITLE
Incorporate mistral.rs into swift MacOS app?

### DISCUSSION THREAD

**Discussion by kinchahoy** [EricLBuehler/mistral.rs#910] [Q&A]
Has anybody put together a set of swift bindings or a best practice guide on how to use mistral.rs inside a MacOS / iOS app? I've seen some reasonable ones for llama.cpp, but they tend to do a poor job of staying current with the package (and llama.cpp has poorer vision LLM support).

**Comment by simoncollins** [comment]
I haven't used it but it seems Mozilla's [uniffi](https://github.com/mozilla/uniffi-rs) might be helpful here. It lets you generate bindings for several languages including [Swift](https://mozilla.github.io/uniffi-rs/latest/swift/overview.html). It appears you need to either create an IDL for your interface or annotate the code with macros.

If you went this approach you'd probably want to create a Rust shim library with your application specific API and generate the bindings for that unless you need the full API of mistral.rs exposed.

There's a [starter repo](https://github.com/ianthetechie/uniffi-starter) for iOS/Android which I assume will generalise to MacOS.

There's also a [cargo plugin](https://github.com/antoniusnaumann/cargo-swift) to create a Swift package that wraps your library.
